### PR TITLE
OAK-10715 : embedded verification fixes : it was using the wrong equa…

### PR DESCRIPTION
**Pull request by stefan-egli**
…ls method that always returned true in the given use case, plus added verifyDeletion

### CODE REVIEW

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 1]
```diff
@@ -1714,7 +1736,48 @@ public void removeGarbage(final VersionGCStats stats) {
...
+                    log.error("verify : traversedParent not a DocumentNodeState : {}",
+                            traversedParent.getClass());
+                    return false;
+                }
+                lastRev = ((DocumentNodeState) traversedParent).getLastRevision();
+            }
+            final NodeState actual = newDoc.getNodeAtRevision(nodeStore, lastRev, lastRevision);
+            // use more thorough version of equals to ensure properties are checked
+            // (the faster state.equals() would stop if lastRev matches,
+            // but as we're fiddling with immuability rule of a document,
```
Comment (Joscorbe) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]:
```suggestion
            // but as we're fiddling with immutability rule of a document,
```

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 2]
```diff
@@ -1639,16 +1640,37 @@ public void removeGarbage(final VersionGCStats stats) {
...
                     };
+                    for (Entry<String, Long> e : orphanOrDeletedRemovalMap.entrySet()) {
+                        NodeState traversedState = root;
+                        for (String name : Path.fromString(Utils.getPathFromId(e.getKey())).elements()) {
+                            traversedState = traversedState.getChildNode(name);
+                        }
+                        if (!verifyDeletion(traversedState)) {
+                            // verification failure
+                            // let's skip this document
+                            it.remove();
```
Comment (Joscorbe) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1678]: Same here

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 3]
```diff
@@ -1639,16 +1640,37 @@ public void removeGarbage(final VersionGCStats stats) {
...
+                            traversedParent = traversedState;
+                            traversedState = traversedState.getChildNode(name);
+                        }
+
                         NodeDocument newDoc = Collection.NODES.newDocument(ds);
                         oldDoc.deepCopy(newDoc);
                         UpdateUtils.applyChanges(newDoc, update);
-                        if (!verify(oldDoc, newDoc, update)) {
+                        // for the time being, verify both with classic and traversed
+                        if (!verifyViaTraversedState(traversedState, traversedParent,
```
Comment (Joscorbe) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1654]: Should we maybe add a debug line here? We may have interest in detecting these cases. Wdyt?

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1654]: +1, added https://github.com/apache/jackrabbit-oak/pull/1377/commits/5b668577a5712f48fa569d998e99ae3fc302c362

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 2 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1678]: +1, added https://github.com/apache/jackrabbit-oak/pull/1377/commits/5b668577a5712f48fa569d998e99ae3fc302c362

**Review by rishabhdaim — APPROVED** [review]

**Pull request merged**